oauthproxy.go

```diff
@@ -820,68 +825,72 @@ func (p *OAuthProxy) getAuthenticatedSession(rw http.ResponseWriter, req *http.R
 func (p *OAuthProxy) addHeadersForProxying(rw http.ResponseWriter, req *http.Request, session *sessionsapi.SessionState) {
 	if p.PassBasicAuth {
 		req.SetBasicAuth(session.User, p.BasicAuthPassword)
-		req.Header["X-Forwarded-User"] = []string{session.User}
-		if session.Email != "" {
-			req.Header["X-Forwarded-Email"] = []string{session.Email}
-		} else {
-			req.Header.Del("X-Forwarded-Email")
-		}
 	}
+	headers, err := p.provider.HeadersToInject(session)
 
-	if p.PassUserHeaders {
-		req.Header["X-Forwarded-User"] = []string{session.User}
-		if session.Email != "" {
-			req.Header["X-Forwarded-Email"] = []string{session.Email}
-		} else {
-			req.Header.Del("X-Forwarded-Email")
-		}
-	}
-
-	if p.SetXAuthRequest {
-		rw.Header().Set("X-Auth-Request-User", session.User)
-		if session.Email != "" {
-			rw.Header().Set("X-Auth-Request-Email", session.Email)
-		} else {
-			rw.Header().Del("X-Auth-Request-Email")
+	if err != nil {
+		log.Printf("Unable to inject headers: %v", err)
+	} else {
+		if p.PassUserHeaders {
+			_, exportEverything := p.exportClaims["*"]
+			for k, vv := range *headers {
+				rename, ok := p.exportClaims[strings.ToLower(k)]
+				if !ok && !exportEverything {
+					continue
+				}
+				headerName := "X-Forwarded-" + k
+				if rename != "" {
+					headerName = "X-Forwarded-" + rename
+				}
+				req.Header.Del(headerName)
+				for i := range vv {
+					req.Header.Add(headerName, vv[i])
+				}
+			}
 		}
-
-		if p.PassAccessToken {
-			if session.AccessToken != "" {
-				rw.Header().Set("X-Auth-Request-Access-Token", session.AccessToken)
-			} else {
-				rw.Header().Del("X-Auth-Request-Access-Token")
+		if p.SetXAuthRequest {
+			_, exportEverything := p.exportClaims["*"]
+			for k, vv := range *headers {
+				rename, ok := p.exportClaims[strings.ToLower(k)]
+				if !ok && !exportEverything {
+					continue
+				}
+				headerName := "X-Auth-Request-" + k
+				if rename != "" {
+					headerName = "X-Auth-Request-" + rename
+				}
+				rw.Header().Del(headerName)
+				for i := range vv {
+					rw.Header().Add(headerName, vv[i])
+				}
 			}
 		}
 	}
-
 	if p.PassAccessToken {
 		if session.AccessToken != "" {
 			req.Header["X-Forwarded-Access-Token"] = []string{session.AccessToken}
+			rw.Header().Set("X-Auth-Request-Access-Token", session.AccessToken)
+
 		} else {
 			req.Header.Del("X-Forwarded-Access-Token")
+			rw.Header().Del("X-Auth-Request-Access-Token")
 		}
 	}
 
-	if p.PassAuthorization {
+	if p.PassAuthorization || p.SetAuthorization {
```

Comment:
I don't think merging these two actually does what is desired?

There are two sets of flags, pass and set flags, the pass ones are typically for passing values to upstream requests, where set ones are for setting them on the response.

I think merging these is not maintaining existing behaviour right?

Reply:
> I think merging these is not maintaining existing behaviour right?

You are right, this should be reverted. I couldn't think of a situation where in I'd be using the proxy in both x-auth-request form AND as a proxy. I still cannot, but I don't think it's an appropriate change in this PR. I was trying to make setting/adding headers a bit easier to follow.